include/boost/geometry/algorithms/similarity.hpp

```diff
@@ -0,0 +1,330 @@
+void reverse_projections(Geometry2 const& target, Projections& projections)
+{
+    std::size_t const last_segment_index = target.size() - 2;
+    for (auto& projection : projections)
+    {
+        auto const si = projection.segment_index;
+        if (si > last_segment_index)
+        {
+            // Defensive check.
+            continue;
+        }
+        auto const segment_length = geometry::distance(target[si], target[si + 1]);
+        projection.segment_index = last_segment_index - si;
+        projection.offset = segment_length - projection.offset;
+    }
+}
+
+// Add "self" points on collection which was created projecting an other line
+// onto this line.
+// Sort all points along the line.
+template <typename It, typename Enriched>
+void enrich_with_self(It begin, It end, Enriched& enriched)
+{
+    using info_t = std::decay_t<decltype(enriched.front())>;
+    std::size_t index = 0;
+    for (It it = begin; it != end; ++it)
+    {
+        info_t info;
+        info.is_projection = false;
+        info.source_index = index;
+        info.segment_index = index;
+        info.point = *it;
+        enriched.push_back(std::move(info));
+        index++;
+    }
+
+    std::sort(enriched.begin(), enriched.end(), 
+        [](const auto& lhs, const auto& rhs){
+            auto const lt = std::tie(lhs.segment_index, lhs.offset, lhs.source_index, lhs.is_projection);
+            auto const rt = std::tie(rhs.segment_index, rhs.offset, rhs.source_index, rhs.is_projection);
+            return lt < rt;
+        });
+}
+
+template <typename Projections>
+bool is_reversed(Projections const& projections)
+{
+    std::size_t count_right_order = 0;
+    std::size_t count_wrong_order = 0;
+    for (std::size_t i = 0; i + 1 < projections.size(); i++)
+    {
+        auto const current = std::tie(projections[i].segment_index, projections[i].offset);
+        auto const next = std::tie(projections[i + 1].segment_index, projections[i + 1].offset);
+
+        if (current <= next)
+        {
+            count_right_order++;
+        }
+        else
+        {
+            count_wrong_order++;
+        }
+    }
+
+    return count_wrong_order > count_right_order;
+}
+
+template <typename Projection, typename Ring>
+bool fill_quadrilateral(Projection const& p0, Projection const& p1, Projection const& q0, Projection const& q1, Ring& ring)
+{
+    using point_t = typename point_type<Ring>::type;
+    using side_strategy = typename strategy::side::services::default_strategy
+    <
+        typename cs_tag<point_t>::type
+    >::type;
+
+    int const side_q0_p = side_strategy::apply(p0.point, p1.point, q0.point);
+    int const side_q1_p = side_strategy::apply(p0.point, p1.point, q1.point);
+    int const side_p0_q = side_strategy::apply(q0.point, q1.point, p0.point);
+    int const side_p1_q = side_strategy::apply(q0.point, q1.point, p1.point);
+
+    if (side_q0_p == 0 && side_q1_p == 0 && side_p0_q == 0 && side_p1_q == 0)
```

Comment:
shouldn't `side_q0_p == 0 && side_q1_p == 0` be enough?

Reply:
That should indeed be enough, at least for the cases I tested. However, I'm not sure about FP-precision where there are sometimes weird results (such as 3 times 0 and one non-zero). And the sides have to be calculated later (used later) anyway.
But with those FP-precision errors the area is probably minimal - so it probably won't make a real difference.

I will think about what my final preference is here.